### PR TITLE
ARCO-204: add cache to arc

### PR DESCRIPTION
**Pull request by pawellewandowski98**
## Description of Changes

Added cache store interface and implementation of redid and freecache

## Linked Issues / Tickets

ARCO-204

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have updated `CHANGELOG.md` with my changes

### CODE REVIEW

File: config/example_config.yaml [hunk 1]
```diff
@@ -32,6 +32,11 @@ peers: # list of bitcoin node peers to connect to
     port:
       p2p: 18335
 
+cache:
+  engine: freecache # cache engine - freecache/redis
+  freecache: # freecache configuration
+    size: 1000000 # size of cache
```
Comment (boecklim) [inline config/example_config.yaml:38]: There are some settings missing

```go
type RedisConfig struct {
	Addr     string `mapstructure:"addr"`
	Password string `mapstructure:"password"`
	DB       int    `mapstructure:"db"`
}
```

Comment (pawellewandowski98) [reply config/example_config.yaml:38]: Added

File: internal/cache/factory.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package cache
+
+import (
+	"errors"
+	"github.com/bitcoin-sv/arc/config"
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/cache/factory.go]: Moved to `cmd/arc/services`

File: internal/cache/freecache_test.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"testing"
+	"time"
+
+	"github.com/bitcoin-sv/arc/internal/cache"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFreecacheStore(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/cache/freecache_test.go]: Removed

File: internal/cache/freecache.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/coocood/freecache"
+)
+
+// FreecacheStore is an implementation of CacheStore using freecache.
+type FreecacheStore struct {
+	cache *freecache.Cache
+}
+
+// NewFreecacheStore initializes a FreecacheStore.
+func NewFreecacheStore(size int) *FreecacheStore {
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/cache/freecache.go]: Done

File: internal/cache/redis.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+// RedisStore is an implementation of CacheStore using Redis.
+type RedisStore struct {
+	client redis.UniversalClient
+	ctx    context.Context
+}
+
+// NewRedisStore initializes a RedisStore.
+func NewRedisStore(addr, password string, db int) *RedisStore {
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/cache/redis.go]: Done

File: internal/cache/redis_test.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+func setupMockRedisStore() (*cache.RedisStore, redismock.ClientMock) {
+	client, mock := redismock.NewClientMock()
+	store := cache.NewRedisStore("", "", 0)
+	store.SetClient(client)
+	store.SetContext(context.Background())
+	return store, mock
+}
+
+func TestRedisStore(t *testing.T) {
+	store, mock := setupMockRedisStore()
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply internal/cache/redis_test.go]: Removed

**Review by arkadiuszos4chain — APPROVED** [review]

**Pull request merged**